hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/transaction/TestLocalProcessLockProvider.java

```diff
@@ -0,0 +1,253 @@
+    final AtomicBoolean writer2Completed = new AtomicBoolean(false);
+
+    // Main test thread
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.lock();
+    });
+
+    // Another writer thread in parallel, should block
+    // and later acquire the lock once it is released
+    Thread writer2 = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        assertDoesNotThrow(() -> {
+          localProcessLockProvider.lock();
+        });
+        assertDoesNotThrow(() -> {
+          localProcessLockProvider.unlock();
+        });
+        writer2Completed.set(true);
+      }
+    });
+    writer2.start();
+
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.unlock();
+    });
+
+    try {
+      writer2.join();
+    } catch (InterruptedException e) {
+      //
+    }
+    Assertions.assertTrue(writer2Completed.get());
+  }
+
+  @Test
+  public void testTryLockAcquisition() {
+    LocalProcessLockProvider localProcessLockProvider = new LocalProcessLockProvider(lockConfiguration, hadoopConfiguration);
+    Assertions.assertTrue(localProcessLockProvider.tryLock());
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.unlock();
+    });
+  }
+
+  @Test
+  public void testTryLockAcquisitionWithTimeout() {
+    LocalProcessLockProvider localProcessLockProvider = new LocalProcessLockProvider(lockConfiguration, hadoopConfiguration);
+    Assertions.assertTrue(localProcessLockProvider.tryLock(1, TimeUnit.MILLISECONDS));
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.unlock();
+    });
+  }
+
+  @Test
+  public void testTryLockReAcquisitionBySameThread() {
+    LocalProcessLockProvider localProcessLockProvider = new LocalProcessLockProvider(lockConfiguration, hadoopConfiguration);
+    Assertions.assertTrue(localProcessLockProvider.tryLock());
+    assertThrows(HoodieLockException.class, () -> {
+      localProcessLockProvider.tryLock(1, TimeUnit.MILLISECONDS);
+    });
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.unlock();
+    });
+  }
+
+  @Test
+  public void testTryLockReAcquisitionByDifferentThread() {
+    LocalProcessLockProvider localProcessLockProvider = new LocalProcessLockProvider(lockConfiguration, hadoopConfiguration);
+    final AtomicBoolean writer2Completed = new AtomicBoolean(false);
+
+    // Main test thread
+    Assertions.assertTrue(localProcessLockProvider.tryLock());
+
+    // Another writer thread
+    Thread writer2 = new Thread(() -> {
+      Assertions.assertFalse(localProcessLockProvider.tryLock(100L, TimeUnit.MILLISECONDS));
+      writer2Completed.set(true);
+    });
+    writer2.start();
+    try {
+      writer2.join();
+    } catch (InterruptedException e) {
+      //
+    }
+
+    assertDoesNotThrow(() -> {
+      localProcessLockProvider.unlock();
+    });
+    Assertions.assertTrue(writer2Completed.get());
```

Comment:
We should actually assert that before we unlock

Reply:
fixed.